crates/ra_hir/src/nameres.rs

```diff
@@ -153,28 +153,103 @@ impl ModuleScope {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ModuleDefOrMacroNs {
+    ModuleDefNs(PerNs<ModuleDef>),
+    MacroNs(MacroDef),
+}
+use ModuleDefOrMacroNs::*;
+
+impl Default for ModuleDefOrMacroNs {
+    fn default() -> ModuleDefOrMacroNs {
+        ModuleDefOrMacroNs::none()
+    }
+}
+
+impl ModuleDefOrMacroNs {
+    pub fn none() -> ModuleDefOrMacroNs {
+        ModuleDefNs(PerNs::none())
+    }
+
+    pub fn take_values(self) -> Option<ModuleDef> {
+        match self {
+            ModuleDefNs(m) => m.take_values(),
+            _ => None,
+        }
+    }
+
+    pub fn take_types(self) -> Option<ModuleDef> {
+        match self {
+            ModuleDefNs(m) => m.take_types(),
+            _ => None,
+        }
+    }
+
+    pub fn is_none(&self) -> bool {
+        match self {
+            ModuleDefNs(m) => m.is_none(),
+            MacroNs(_) => false,
+        }
+    }
+
+    pub fn as_module_def(&self) -> Option<&PerNs<ModuleDef>> {
+        match self {
+            ModuleDefNs(m) => Some(m),
+            _ => None,
+        }
+    }
+
+    pub fn map_module_def<U>(self, f: impl Fn(ModuleDef) -> U) -> PerNs<U> {
+        match self {
+            ModuleDefNs(m) => m.map(&f),
+            _ => PerNs::none().map(&f),
+        }
+    }
+
+    pub fn as_macros(&self) -> Option<&MacroDef> {
+        match self {
+            MacroNs(m) => Some(m),
+            _ => None,
+        }
+    }
+
+    pub fn or(self, other: ModuleDefOrMacroNs) -> ModuleDefOrMacroNs {
+        match (self, other) {
+            (ModuleDefNs(s), ModuleDefNs(o)) => ModuleDefNs(s.or(o)),
+            (MacroNs(s), _) => MacroNs(s),
+            (ModuleDefNs(s), MacroNs(o)) => {
+                if !s.is_none() {
+                    ModuleDefNs(s)
+                } else {
+                    MacroNs(o)
+                }
+            }
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, Eq, Default)]
 pub struct Resolution {
     /// None for unresolved
-    pub def: PerNs<ModuleDef>,
+    pub def: ModuleDefOrMacroNs,
     /// ident by which this is imported into local scope.
     pub import: Option<ImportId>,
 }
 
 #[derive(Debug, Clone)]
 struct ResolvePathResult {
-    resolved_def: PerNs<ModuleDef>,
```

Comment:
This I think should stay an `PerNs`. Partial resolutions make sense when we resolve associated items, and we shouldn't deal with macros there

Reply:
But what happens if it is not a partial resolution ? For example in `resolve_name_in_crate_root_or_extern_prelude` , it is using the scope items directly, which means there is a chance it already fully resolved. Or do i misunderstand something?

Reply:
I think this is exactly the place were we want to resovler methods, one with partial resolution which doesn't deal with macros, another with full and macro-aware resolution.

Need to take a closer look at the code to answer exactly

Reply:
> I think this is exactly the place were we want to resovler methods, one with partial resolution which doesn't deal with macros, another with full and macro-aware resolution.

Oh... I understand what you mean now. :)

Reply:
I try to resolve this problem by making `ModuleDefOrMacroNs` private. So normal resolution do not deal with macros. It is because only `collector.rs` have to be aware about macro and its implementation detail of item resolution.